### DISCUSSION TITLE
Automate npm publishing for WordPress packages

### DISCUSSION THREAD

**Discussion by gziolo** [WordPress/gutenberg#37820] [Developer Experience]
We have GitHub automation in place for handling the [Gutenberg plugin releases](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#plugin-releases).

Prior work:
- https://github.com/WordPress/gutenberg/pull/27488
- https://github.com/WordPress/gutenberg/pull/28138

It would be great to bring the same experience when publishing WordPress packages to npm. It's still a partially manual process that has its quirks when the WordPress trunk branch can be closed or in "feature-freeze" mode as documented in [Synchronizing WordPress Trunk](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#synchronizing-wordpress-trunk).

Let's discuss:
- How to further automate npm publishing steps like backporting commits created by Lerna to the `trunk` or change the flow in a way where we could auto-approve all CLI prompts?
- Ways to modify the existing publishing cadence to make it predictable, for example - fully tied to Gutenberg plugin releases.
- What needs to happen to start using [npm access tokens](https://docs.npmjs.com/creating-and-viewing-access-tokens) to publish packages to npm from CI?

**Comment by ockham** [comment]
Draft PR: https://github.com/WordPress/gutenberg/pull/37751

**Comment by gziolo** [comment]
The current schedule for npm publishing looks as follows:

-   [Synchronizing WordPress Trunk](#synchronizing-wordpress-trunk) (`latest` dist tag) – **when there is no "feature-freeze" mode** in WordPress Core, publishing happens **every two weeks** based on the newly created RC1 version of the Gutenberg plugin. Otherwise, only bug fixes get manually included and published to npm before every next beta and RC version of the following WordPress release.
-   [Minor WordPress Releases](#minor-wordpress-releases) (`patch` dist tag) – only **when bug fixes or security releases** need to be backported into WordPress Core.
-   [Development Releases](#development-releases) (`next` dist tag) – **when there is a "feature-freeze" mode** in WordPress Core, publishing can still happen **every two weeks** based on the new RC1 version of the Gutenberg plugin. It is also possible to perform development releases when there is a need to test the upcoming changes.

It would be great if we could develop a revised strategy where the "feature-freeze" mode in WordPress Core doesn't have any impact on when and how WordPress packages are published to npm. It might require a few changes in how we handle the dist-tags, but maybe we should exercise extreme cases to find something in the middle that works in most cases.

The other issue discussed in https://github.com/WordPress/gutenberg/issues/35630 is the lack of identifying npm releases tied to a specific WordPress major version or the Gutenberg plugin version. It would be beneficial for plugin authors to be able to install npm packages pinned to those configurations when willing to support ranges of plugins/WordPress core versions. At the moment, we don't even publish npm packages for every Gutenberg plugin release so we might want to address that as well.

**Comment by ockham** [comment]
> - How to further automate npm publishing steps like backporting commits created by Lerna to the `trunk`

For context, the existing script currently prompts the user to do so manually after it finishes:
https://github.com/WordPress/gutenberg/blob/29ec9e0261c1c915bb75db4c4679f91225f685d1/bin/plugin/commands/packages.js#L418-L424

I think that this part is the easiest to automate; we already have precedent in the build-plugin workflow: https://github.com/WordPress/gutenberg/blob/29ec9e0261c1c915bb75db4c4679f91225f685d1/.github/workflows/build-plugin-zip.yml#L141-L151 

The main task here will be to figure out the correct SHA(s) produced by `lerna`, but I don't think that should be too hard.

**Reply by gziolo** [reply]
Yes, the only challenge is that one of the commits is optional, but they always have the same title so we can use that for matching.

**Reply by gziolo** [reply]
I sorted out cherry-picking all necessary commits to the `trunk` branch in #38993.

**Comment by ockham** [comment]
> or change the flow in a way where we could auto-approve all CLI prompts?

As for the ones that are part of the `package.js` script that we're currently using to publish npms, my current workaround in #37751 is to translate that JS file to Shell Script (and inline them into the GHA workflow YAML file) -- without the prompts 😬 Alternatively, we could modify `package.js` to only prompt if it wasn't given a `--ci` flag.

**Reply by gziolo** [reply]
Right, it's actually one function `askForConfirmation` that is used for prompts in the Node.js tool so it might be not that hard to ignore those questions.  There is also a CLI option for Lerna to skip all prompts `--yes`. See https://github.com/lerna/lerna/tree/main/commands/publish#--yes and https://github.com/lerna/lerna/tree/main/commands/version#--yes.

Still, there are some steps that are obsolete on CI – you don't need to clone the repository into a temp directory and take care of the cleanup.

**Reply by gziolo** [reply]
I added `--ci` flag to the npm publishing script with https://github.com/WordPress/gutenberg/pull/38993 and https://github.com/WordPress/gutenberg/pull/39199. I was able to successfully publish to npm with no CLI interactions.

**Comment by ockham** [comment]
> - Ways to modify the existing publishing cadence to make it predictable, for example - fully tied to Gutenberg plugin releases.

I'm leaning towards changing the workflow trigger from manual (which I've currently put for testing purposes in #37751) to maybe upon release (which is also the trigger for the ["Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo" workflow](https://github.com/WordPress/gutenberg/blob/29ec9e0261c1c915bb75db4c4679f91225f685d1/.github/workflows/upload-release-to-plugin-repo.yml)) -- and limit it to publishing RC1, to replicate our currently documented recommendation.

**Reply by gziolo** [reply]
I like that. Having a way to trigger the release on-demand and tied to the Gutenberg plugin RC1 seems perfect.

I'd also like to explain why we switched to RC1 as the source for npm publishing. If the publishing happens just after triggering the release process for the plugin, we remove the risk that the changelog files will have merge conflicts when backporting them back to `trunk.` It also ensures some of the new commits get accidentally attributed to one of the already published versions since there is no new version in the CHANGELOG at the time of merging. This is coming from the fact that we keep items under 'Unreleased' in the release branch and in `trunk` until the npm publishing starts.

We could also rethink the way CHANGELOGs are maintained but let's keep it out of the scope for now.

**Reply by gziolo** [reply]
I'm experimenting with triggering npm publishing at the time when RC1 for the Gutenberg plugin is released. See #39259.

**Comment by ockham** [comment]
> - What needs to happen to start using [npm access tokens](https://docs.npmjs.com/creating-and-viewing-access-tokens) to publish packages to npm from CI?

This might be the trickiest one, especially in a lerna context. I'm hoping that we'll be able to find a solution that works by "just" storing some kind of token in GH Secrets.

**Reply by gziolo** [reply]
> Automation: an automation token can download packages and publish new ones, but if you have two-factor authentication (2FA) configured on your account, it will not be enforced.

It looks like it bypasses 2FA when using the token from CI. 

I also found an example of how it can be configured on CI in https://github.com/semantic-release/npm/issues/277#issuecomment-787128568:

```yaml
- name: Publish
  run: |
    echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
    npm publish --tag ${{ steps.deployment.outputs.tag }}
  env:
    NPM_TOKEN: ${{ secrets.NPM_DEPLOY_KEY }}
```

**Reply by gziolo** [reply]
I found this excellent article https://dev.to/astagi/publish-to-npm-using-github-actions-23fn which contains all the steps required to set up the token and use it in a GitHub action. I'm experimenting with it in #39259.

**Comment by ockham** [comment]
> It would be great if we could develop a revised strategy where the "feature-freeze" mode in WordPress Core doesn't have any impact on when and how WordPress packages are published to npm. It might require a few changes in how we handle the dist-tags, but maybe we should exercise extreme cases to find something in the middle that works in most cases.

Yeah, we can try to figure out something like that. OTOH, as a first step for automation, I like to replicate the exact behavior of the current script in CI, since there's always a bit of a risk of feature creep (especially if some changes need to be agreed upon -- easier to tackle those in follow-up PRs). Maybe there's some way to automatically determine (through some REST API maybe?) if WP Core is currently in feature-freeze mode.

**Reply by gziolo** [reply]
For the sake of initial implementation, we can completely ignore the struggle with detecting the feature freeze or when a minor version bump has to happen to account for the next major WordPress version. We can focus on the development releases with the `next` dist-tag. It also removes any risk that we break the npm packages published to npm with the default `latest` dist-tag.

**Comment by ockham** [comment]
> The other issue discussed in #35630 is the lack of identifying npm releases tied to a specific WordPress major version or the Gutenberg plugin version. It would be beneficial for plugin authors to be able to install npm packages pinned to those configurations when willing to support ranges of plugins/WordPress core versions. At the moment, we don't even publish npm packages for every Gutenberg plugin release so we might want to address that as well.

This makes sense but could be a bit tricky to do, especially if we want our npm releases to follow semver. I think this is best discussed separately from the goal of automating the npm release process :)

**Reply by gziolo** [reply]
What if we get a bit creative and use `[npm-dist-tag]` as proposed in https://github.com/WordPress/gutenberg/issues/24376 by @sirreal?

It should be straightforward in the case of Gutenberg plugin releases. After publishing packages to npm, you can read all production packages and run the following command using the extracted package name and version:

```bash
npm dist-tag add <pkg>@<version> [<tag>]
```

**Comment by gziolo** [comment]
> It would be great if we could develop a revised strategy where the "feature-freeze" mode in WordPress Core doesn't have any impact on when and how WordPress packages are published to npm. It might require a few changes in how we handle the dist-tags, but maybe we should exercise extreme cases to find something in the middle that works in most cases.

I'm proposing a new strategy in #39259 where the "feature-freeze" mode in WordPress Core isn't taken into account when publishing npm packages with the `latest` tag. The plan is to publish to npm every time the Gutenberg plugin RC1 is released. The only remaining question is whether we would use `patch` dist-tag for beta and RC phase of major WordPress releases or we switch to tags targeting a given release line as discussed in https://github.com/WordPress/gutenberg/discussions/37820#discussioncomment-1961128. In the case of the upcoming WordPress 6.0 release, we could start using `wp-6.0` tag instead.

**Reply by youknowriad** [reply]
I'm fine with changes, just want to mention two things:

 - During the beta/rc cycle of WordPress, we often need to make package updates that are not necessary just bug fixes, so minor version changes, we should make sure there's no "newer" version available because of a plugin npm release.
 - After a release of WordPress, we need to leave room for patch releases and security releases (so patch versions changes).

**Reply by gziolo** [reply]
> After a release of WordPress, we need to leave room for patch releases and security releases (so patch versions changes).

Every plugin release (RC1) will trigger at least a `minor` version bump for packages with production code. I also want to automate adding a CHANGELOG entry that links to the GitHub release page with the full plugin release changelog.

> During the beta/rc cycle of WordPress, we often need to make package updates that are not necessary just bug fixes, so minor version changes, we should make sure there's no "newer" version available because of a plugin npm release.

As long as those changes are applied as a `patch` version bump then it should be good. An alternative would be to enforce a `major` version bump to production packages on the plugin release that won't be included in the upcoming major WordPress release. The latter is more difficult to handle but I think @ockham had some ideas on how to detect this exact moment.

**Comment by gziolo** [comment]
I was thinking about the changes to the publishing process based on the feedback in #39259 from @tellthemachines and this is my current thinking.

<h2>Proposed revised strategy for npm publishing</h2>

The Gutenberg repository follows the [WordPress SVN repository](https://make.wordpress.org/core/handbook/about/release-cycle/) in the branching strategy for every major WordPress release. In addition to that, it also contains two other special branches that control npm publishing workflows:

<ul><li>The <code>wp/latest</code> <em>(instead of <code>wp/trunk</code> as currently</em>) branch contains the same version of packages published to npm with the <code>latest</code> distribution tag. The goal here is to have this branch synchronized with the last Gutenberg plugin release unless there is a hotfix release required.</li><li>The <code>wp/next</code> branch contains the same version of packages published to npm with the <code>next</code> distribution tag. It's always synchronized with the <code>trunk</code> branch <em>(rather than the plugin branch as now)</em>. Projects should use those packages for development or testing purposes only.</li><li>A Gutenberg branch targeting a specific WordPress major release (including its minor version increments) is created (example <code>wp/5.2</code>) based on the <code>wp/latest</code> Gutenberg branch just after the last Gutenberg release planned for inclusion in the next major WordPress release. </li></ul>

<p>Release types and their schedule:</p>

<ul><li><a href="https://github.com/WordPress/gutenberg/pull/39259/files#synchronizing-gutenberg-plugin">Synchronizing Gutenberg Plugin)</a>&nbsp;(<code>latest</code>&nbsp;dist tag) – publishing happens every two weeks based on the newly created RC1 version of the Gutenberg plugin.</li><li><a href="https://github.com/WordPress/gutenberg/pull/39259/files#wordpress-releases">WordPress Releases</a>&nbsp;(<code>patch</code>&nbsp;dist tag) – only when bug or security fixes need to be backported into WordPress Core. It applies both to the older major versions of WordPress and the current WordPress major version release when it reaches beta and RC phases.</li><li><a href="https://github.com/WordPress/gutenberg/pull/39259/files#development-releases">Development Releases</a>&nbsp;(<code>next</code>&nbsp;dist tag) – it is also possible to perform development releases at any time when there is a need to test the upcoming changes.</li></ul>

**Reply by youknowriad** [reply]
Maybe I'm misunderstanding something but In this proposal, it's unclear which packages to use for the first beta of a major WordPress release and how we go about that?

**Reply by ockham** [reply]
If we're mapping the individual release types to automatic triggers (upon RC publishing) vs manual ones (i.e. [`workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch)), then:

> * [Synchronizing Gutenberg Plugin)](https://github.com/WordPress/gutenberg/pull/39259/files#synchronizing-gutenberg-plugin) (latest dist tag) – publishing happens every two weeks based on the newly created RC1 version of the Gutenberg plugin.

... would be automatic, and

> * [WordPress Releases](https://github.com/WordPress/gutenberg/pull/39259/files#wordpress-releases) (patch dist tag) – only when bug or security fixes need to be backported into WordPress Core. It applies both to the older major versions of WordPress and the current WordPress major version release when it reaches beta and RC phases.
> * [Development Releases](https://github.com/WordPress/gutenberg/pull/39259/files#development-releases) (next dist tag) – it is also possible to perform development releases at any time when there is a need to test the upcoming changes.

... would be manual -- likely with some input field to specify whether to do a `patch` or a `next` release. (#37751 already has a bit of that logic.)

**Reply by youknowriad** [reply]
> ... would be automatic

The comment above suggest using wp/latest (and not wp/trunk) for the releases. how do we update wp/trunk in that case? What's Automatic here is releasing packages from Gutenberg releases, not updating WordPress core trunk with these packages AFAIK

**Reply by gziolo** [reply]
> The comment above suggest using wp/latest (and not wp/trunk) for the releases. how do we update wp/trunk in that case? 

In the automated workflow, `wp/latest` better describes its role, so I think we can rename `wp/trunk` to `wp/latest`. Otherwise, it would be confusing when npm packages contain changes that won't get included in the ongoing WP major release during beta/RC phase.

> It's unclear which packages to use for the first beta of a major WordPress release and how we go about that?

It's very important to describe it properly because it's very confusing. My current understanding is that we use `wp/latest` to synchronize changes with WordPress core, including the last Gutenberg plugin release that gets included in the current major WordPress release. Once we reach that point where we know we will only cherry-pick commits from the Gutenberg repository, we create `wp/*` (example: `wp/6.0`) branch and use it for npm publishing with the `patch` dist-tag and use for synchronization with WordPress core.

**Reply by youknowriad** [reply]
>Once we reach that point where we know we will only cherry-pick commits from the Gutenberg repository, we create wp/*

Thanks for explaining, this works. It creates a difference between WP branches and Gutenberg wp/* branches (not fully synced) but that's fine I think.

**Reply by gziolo** [reply]
>  It creates a difference between WP branches and Gutenberg wp/* branches (not fully synced) but that's fine I think.

Yes, it's a bit unfortunate, but at the same time it better aligns with the process we have been doing for such a long time. @tellthemachines correctly pointed out that `wp/trunk` is confusing in the period where for WordPress core you use `trunk` to commit everything but we are already past that in the Gutenberg plugin and we cherry-pick individual commits for syncing. I guess I need to explain it better in the documentation proposed.

**Reply by tellthemachines** [reply]
> It creates a difference between WP branches and Gutenberg wp/* branches (not fully synced)

We have very different workflows and release cadence between Core and Gutenberg, so I think it's fine to have that difference as long as it's clearly documented. And hopefully at some point Core will evolve to creating the release branch straight after Beta 😄

**Reply by gziolo** [reply]
I updated https://github.com/WordPress/gutenberg/pull/39389 with all the changes discussed above. I hope I was able to cover all the discussion points and it's easier to follow now.

**Comment by gziolo** [comment]
I opened https://github.com/WordPress/gutenberg/pull/40927 which automates the process of publishing npm packages targeting the given major WordPress release.

**Comment by gziolo** [comment]
The remaining tasks I can think of at the moment:

- [x] ~set an additional `gb-X.Y` dist-tag when publishing packages based on the Gutenberg release~ - we would have new dist-tags added every two weeks, we had over 130 plugin releases, so maybe it's too much to maintain
- [x] run and document how the new workflow works for manual publishing
- [ ] ensure that the release type `wp` works correctly with older `wp/5.x` branches that use Node 10 or 12
- [x] update documentation for Block Editor Release Process for Major Releases - at least the [Managing Weekly Beta and RC releases](https://make.wordpress.org/core/handbook/about/release-cycle/block-editor-release-process-for-major-releases/#managing-weekly-beta-and-rc-releases)
- [x] post at https://make.wordpress.org/core/ with the summary of the changes applied to npm publishing
  - new npm user for publishing https://www.npmjs.com/~gutenbergplugin and approval from the Gutenberg core team (should we add all core committers or a subset of them?) 
  - automatic npm publishing tied to Gutenberg releases
  - manually triggered npm publishing for WP major releases and the trick with `dist-tag` to help with the flow

**Reply by gziolo** [reply]
I used the following function to apply `wp-X.Y` dis-tag for WordPress releases 5.7, 5.8 and 5.9:
```js
async function addDistTagToPackages( { gitWorkingDirectoryPath } ) {
	const { dependencies, devDependencies } = require( path.resolve(
		gitWorkingDirectoryPath,
		'package.json'
	) );
	const packageJsonFiles = await glob(
		path.resolve( gitWorkingDirectoryPath, 'packages/*/package.json' )
	);

	const operations = packageJsonFiles
		.map( ( packageJsonFile ) => require( packageJsonFile ) )
		.filter(
			( packageJson ) =>
				packageJson.private !== true &&
				( Object.keys( dependencies ).includes( packageJson.name ) ||
					Object.keys( devDependencies ).includes(
						packageJson.name
					) )
		)
		.map(
			( packageJson ) =>
				`npm dist-tag add ${ packageJson.name }@${ packageJson.version } wp-5.7 --otp 123456;`
		);
	let skip = true;
	for ( const operation of operations ) {
		if ( operation.includes( 'jest-preset-default' ) ) {
			skip = false;
		}
		if ( skip ) continue;
		try {
			//console.log( operation );
			await command( operation, {
				cwd: gitWorkingDirectoryPath,
				stdio: 'inherit',
			} );
		} catch ( error ) {}
	}
}
```

We already had WordPress 6.0 covered. With all that it's possible to easily switch between releases WP 5.7-6.0 using `@wordpress/scripts`:

```
npx wp-scripts packages-update --dist-tag=wp-5.8
```
